module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+ * randomly. The first 64 bits of the IV are stored in the second word of DVA[2]
+ * and the remaining 32 bits are stored in the upper 32 bits of blk_fill. For
+ * most object types this is safe because we only encrypt level 0 blocks which
+ * means that the fill count will be 1. For DMU_OT_DNODE blocks the fill count
+ * is instead used to indicate the number of allocated dnodes beneath the bp.
+ * The on-disk format supports at most 2^15 slots per L0 dnode block, because
+ * the maximum block size is 16MB (2^24). In either case, for level 0 blocks
+ * this number will still be smaller than UINT32_MAX so it is safe to store the
+ * IV in the top 32 bits of blk_fill, while leaving the bottom 32 bits of the
+ * fill count for the dnode code.
+ *
+ * Master key:
+ * This is the most important secret data of an encrypted dataset. It is used
+ * along with the salt to generate that actual encryption keys via HKDF. We
+ * do not use the master key to directly encrypt any data because there are
+ * theoretical limits on how much data can actually be safely encrypted with
+ * any encryption mode. The master key is stored encrypted on disk with the
+ * user's wrapping key. It's length is determined by the encryption algorithm.
+ * For details on how this is stored see the block comment in dsl_crypt.c
+ *
+ * Salt:
+ * Used as an input to the HKDF function, along with the master key. We use a
+ * 64 bit salt, stored unencrypted in the first word of DVA[2]. Any given salt
+ * can be used for encrypting many blocks, so we cache the current salt and the
+ * associated derived key in zio_crypt_t so we do not need to derive it again
+ * needlessly.
+ *
+ * Encryption Key:
+ * A secret binary key, generated from an HKDF function used to encrypt and
+ * decrypt data.
+ *
+ * Message Authenication Code (MAC)
+ * The MAC is an output of authenticated encryption modes such as AES-GCM and
+ * AES-CCM. Its purpose is to ensure that an attacker cannot modify encrypted
+ * data on disk and return garbage to the application. Effectively, it is a
+ * checksum that can not be reproduced by an attacker. We store the MAC in the
+ * second 128 bits of blk_cksum, leaving the first 128 bits for a truncated
+ * regular checksum of the ciphertext which can be used for scrubbing.
+ *
+ * OBJECT AUTHENTICATION:
+ * Some object types, such as DMU_OT_MASTER_NODE cannot be encrypted because
+ * they contain some info that always needs to be readable. To prevent this
+ * data from being altered, we authenticate this data using SHA512-HMAC. This
+ * will produce a MAC (similar to the one produced via encryption) which can
+ * be used to verify the object was not modified. HMACs do not require key
+ * rotation or IVs, so we can keep all 3 copies of authenticated data.
+ *
+ * ZIL ENCRYPTION:
+ * ZIL blocks have their bp written to disk ahead of the associated data, so we
+ * cannot store encryption paramaters there as we normally do. For these blocks
```

Comment:
I will rephrase that sentence. Basically I was trying to say we generate the salt and IV at block allocation time and the MAC is stored in the embedded ZIL checksum.